File: apps/web/src/api/share.ts
import { apiClient } from './client'

export interface ShareRequest {
  recipient_email: string
  message?: string
}

export interface ProjectShareRequest extends ShareRequest {
  access_type: 'view' | 'edit'
}

export interface MilestoneShareRequest extends ShareRequest {
  access_type: 'view' | 'edit'
}

export type ShareErrorCode =
  | 'google_not_connected'
  | 'gmail_scope_missing'
  | 'google_token_refresh_failed'

export interface ShareErrorDetail {
  error_code: ShareErrorCode
}

export async function shareProject(projectId: string, data: ProjectShareRequest): Promise<void> {
  await apiClient.post(`/projects/${projectId}/share`, data)
}

export async function shareMilestone(milestoneId: string, data: MilestoneShareRequest): Promise<void> {
  await apiClient.post(`/milestones/${milestoneId}/share`, data)
}

File: apps/api/schemas/share.py
from typing import Literal

from pydantic import BaseModel, EmailStr


class ShareRequest(BaseModel):
    recipient_email: EmailStr
    message: str | None = None


class ProjectShareRequest(ShareRequest):
    access_type: Literal['view', 'edit'] = 'edit'


class MilestoneShareRequest(ShareRequest):
    access_type: Literal['view', 'edit'] = 'view'


class ShareErrorCode(BaseModel):
    error_code: str


class ShareErrorDetail(BaseModel):
    error_code: str


class ShareResponse(BaseModel):
    sent: bool = True
